doc/designs/fast-param-find.md

```diff
@@ -0,0 +1,324 @@
+- "params" defines the parameters both gettable and settable
+
+Within the "params" the fields specify each parameter by label.
+
+Each parameter is then specialised with attibutes:
+
+- "type" is the OSSL_PARAM type
+- "ctype" is the underlying C type (e.g. for an integer parameter size_t could be the C type)
+- "access" is readwrite, readonly or writeonly.  This determines if the parameter is a settable, gettable or both
+- "field" is an accessor to the field itself
+- "sanitycheck" is a validation check for the parameter.  If present, code will be generated `if (!(sanitycheck)) return 0;`
+  The local variable `var` will contain the C value if specified.
+- "setaction" is C code to execute when the parameter is being set.  It will define an OSSL_PARAM pointer p to set.
+- "code" set to "no" skips code generation for this parameter, it defaults to "yes" which generates handlers.  This is useful
+  when a parameter is duplicated with differenting types (e.g. utf8 string and integer).
+- "published" set to "yes" includes the parameter in the gettable/settable lists.  Set to "no" and it isn't included (but will
+  still be processed).  It defaults to "yes".
+
+- Flags include:
+  - nostatic: do not make the function static
+  - nocode: do not generate code for this parameter
+    - This allows, e.g., two different types for a parameter (int & string)
+  - unpublished: do not generate this parameter in the gettable/settable list
+    - "engine" is the only one like this
+  - readonly: create a getter but not a setter
+  - writeonly: create a setting but not a getter
+
+The idea is that the gettable and get functions will be simultaneously generated
+along with fast decoder to look up parameter names quickly.
+
+The getter and setter functions will be pre-populated with some local variable:
+
+    OSSL_PARAM *p;          /* The matching parameter */
+    type val;               /* The value of the parameter after a get/set call (for C types) */
+
+
+A worked example for scrypt:
+
+Would generate something along the lines of:
+
+```c
+enum kdf_scrypt_ctx_param_e {
+    kdf_scrypt_ctx_param_INVALID,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_PASSWORD,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_PROPERTIES,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SALT,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_MAXMEM,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_N,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_P,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SCRYPT_R,
+    kdf_scrypt_ctx_param_OSSL_KDF_PARAM_SIZE
+};
+
+
+static enum kdf_scrypt_ctx_param_e kdf_scrypt_ctx_lookup(const OSSL_PARAM *p) {
+    /* magic decoder */
+    return kdf_scrypt_ctx_param_INVALID;
+}
+
+static int kdf_scrypt_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p;
+    KDF_SCRYPT *ctx = vctx;
+
+    if (params == NULL)
+        return 1;
+
+    for (p = params; p->key != NULL; p++)
```

Comment:
nit: {

Reply:
Doesn't need the { with a single statement body 😀